rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,226 @@
+succeed either and continue to be throttled.
+
+The traditional storage layout also tightly couples the partitions as folders under the table path. However,
+some users want flexibility to be able to distribute files/partitions under multiple different paths across cloud stores,
+hdfs etc. based on their specific needs. For example, customers have use cases to distribute files for each partition under
+a separate S3 bucket with its individual encryption key. It is not possible to implement such use-cases with Hudi currently.
+
+The high level proposal here is to introduce a new storage layout strategy, where all files are distributed evenly across
+multiple randomly generated prefixes under the Amazon S3 bucket, instead of being stored under a common table path/prefix.
+This would help distribute the requests evenly across different prefixes, resulting in Amazon S3 to create partitions for
+the prefixes each with its own request limit. This significantly reduces the possibility of hitting the request limit
+for a specific prefix/partition.
+
+In addition, we want to expose an interface that provides users the flexibility to implement their own strategy for
+distributing files if using the traditional Hive storage layout or federated storage layer (proposed in this RFC) does
+not meet their use-case.
+
+## Design
+
+### Interface
+
+```java
+/**
+ * Interface for providing storage file locations.
+ */
+public interface FederatedStorageStrategy extends Serializable {
+  /**
+   * Return a fully-qualified storage file location for the given filename.
+   *
+   * @param fileName data file name
+   * @return a fully-qualified location URI for a data file
+   */
+  String storageLocation(String fileName);
+
+  /**
+   * Return a fully-qualified storage file location for the given partition and filename.
+   *
+   * @param partitionPath partition path for the file
+   * @param fileName data file name
+   * @return a fully-qualified location URI for a data file
+   */
+  String storageLocation(String partitionPath, String fileName);
+}
+```
+
+### Generating file paths for Cloud storage optimized layout
+
+We want to distribute files evenly across multiple random prefixes, instead of following the traditional Hive storage
+layout of keeping them under a common table path/prefix. In addition to the `Table Path`, for this new layout user will
+configure another `Table Storage Path` under which the actual data files will be distributed. The original `Table Path` will
+be used to maintain the table/partitions Hudi metadata.
+
+For the purpose of this documentation lets assume:
+```
+Table Path => s3://<table_bucket>/<hudi_table_name>/
+
+Table Storage Path => s3://<table_storage_bucket>/
+```
+Note: `Table Storage Path` can be a path in the same Amazon S3 bucket or a different bucket. For best results,
+`Table Storage Path` should be a bucket instead of a prefix under the bucket as it allows for S3 to partition sooner.
```

Comment:
I think we should rephrase this that storing it w/in a folder make tables share the prefix, and storing the table at the top level allows to avoid that